superset-frontend/src/explore/components/DatasourcePanel/index.tsx

```diff
@@ -245,93 +258,121 @@ export default function DataSourcePanel({
     setInputValue('');
   }, [columns, datasource, metrics]);
 
-  const metricSlice = showAllMetrics
-    ? lists.metrics
-    : lists.metrics.slice(0, DEFAULT_MAX_COLUMNS_LENGTH);
-  const columnSlice = showAllColumns
-    ? lists.columns
-    : lists.columns.slice(0, DEFAULT_MAX_METRICS_LENGTH);
+  const metricSlice = useMemo(
+    () =>
+      showAllMetrics
+        ? lists.metrics
+        : lists.metrics.slice(0, DEFAULT_MAX_COLUMNS_LENGTH),
+    [lists.metrics, showAllMetrics],
+  );
+  const columnSlice = useMemo(
+    () =>
+      showAllColumns
+        ? lists.columns
+        : lists.columns.slice(0, DEFAULT_MAX_METRICS_LENGTH),
+    [lists.columns, showAllColumns],
+  );
```

Comment:
Not yours, but shouldn't this be `DEFAULT_MAX_METRICS_LENGTH` on `metricsSlice` and `DEFAULT_MAX_COLUMNS_LENGTH` on `columnSlice`?